evm_loader/test_cli_emulate.py

```diff
@@ -139,96 +165,250 @@ def test_successful_cli_emulate(self):
+                                   + base58.b58decode(self.token_acc1).hex()
+                                   + base58.b58decode(self.token_acc2).hex()
+                                   + "%064x" % (transfer_amount * (10 ** 9))
+                                   + bytes(self.acc.public_key()).hex()
+                                   )
+
+        emulate_result = emulate_external_call(self.ethereum_caller.hex(),
+                                               self.contract.ethereum_id.hex(),
+                                               trx_data.hex())
 
+        self.compare_tmpl_and_emulate_result_2(tmpl_json, emulate_result)
         # no changes after the emulation
-        self.assertEqual(self.spl_token.balance(self.token_acc1), balance1 + mint_amount)
-        self.assertEqual(self.spl_token.balance(self.token_acc2), balance2)
+        self.assertEqual(self.spl_token.balance(self.token_acc1), balance1 + mint_amount - transfer_amount)
+        self.assertEqual(self.spl_token.balance(self.token_acc2), balance2 + transfer_amount)
+
+        solana_accounts = [AccountMeta(pubkey=item['pubkey'],
+                                     is_signer=item['is_signer'],
+                                     is_writable=item['is_writable'])
+                         for item in emulate_result['solana_accounts']]
+
+        print('solana_accounts:', solana_accounts)
+
+        result = self.contract.call(self.ethereum_caller, trx_data, solana_accounts)
+
+        src_data = result['result']['meta']['innerInstructions'][-1]['instructions'][-1]['data']
+        self.assertEqual(base58.b58decode(src_data)[0], 6)  # 6 means OnReturn
+        self.assertLess(base58.b58decode(src_data)[1], 0xd0)  # less 0xd0 - success
+
+        self.assertEqual(self.spl_token.balance(self.token_acc1), balance1 + mint_amount - 2 * transfer_amount)
+        self.assertEqual(self.spl_token.balance(self.token_acc2), balance2 + 2 * transfer_amount)
+
+
+def test_unsuccessful_cli_emulate(self):
+    balance1 = self.spl_token.balance(self.token_acc1)
+    balance2 = self.spl_token.balance(self.token_acc2)
+    mint_amount = 100
+    self.spl_token.mint(self.token, self.token_acc1, mint_amount)
+    self.assertEqual(self.spl_token.balance(self.token_acc1), balance1 + mint_amount)
+    self.assertEqual(self.spl_token.balance(self.token_acc2), balance2)
+
+    transfer_amount = self.spl_token.balance(self.token_acc1) + 1
+    (trx_data, result) \
+        = self.contract.transfer_ext(self.ethereum_caller,
+                                     self.token, self.token_acc1, self.token_acc2, transfer_amount * (10 ** 9),
+                                     self.acc.public_key()._key)
+    self.assertEqual(result, None)
+
+    emulate_result = emulate_external_call(self.ethereum_caller.hex(),
+                                           self.contract.ethereum_id.hex(),
+                                           trx_data.hex())
```

Comment:
Why do you need these actions in the test? As I understand it, they don't add specific actions for the test.

Reply:
The test **test_unsuccessful_cli_emulate** now can check only by 
self.assertEqual(result, None)
The emulation of the negative scenario now don't work correctly. It will be changed by the #74